### ISSUE TITLE
Twig Extension no longer rendering output

### ISSUE THREAD

**Issue by tacman** [twigphp/Twig#4110]
Did something change in user-defined twig extensions?  This is in the category of "I swear this used to work!"  My extension is at https://github.com/tacman/twig-tree-tag.

It's a slick way of making recursive calls to tree elements, like menus
```twig
    {% set food = [
        {   name: 'fruits', 
            children: [{name: 'apple'},{name:'banana'}]        },
        {   name: 'veggies',
            children: [
            {name: 'peas'},
            {name: 'carrots'}
        ]
        }
    ] %}
    <h1>Food Tree</h1>
    {% tree item in food %}
        {% if treeloop.first %}<ul>{% endif %}
        <li>
            {{ item.name }}
            {% subtree item.children|default([]) %}
        </li>
        {% if treeloop.last %}</ul>{% endif %}
    {% endtree %}

{% endblock %}
```

It used to display the nested HTML, and now nothing.  It's registered okay, but no content either.  I've created a repo showing the bug

```bash
git clone git@github.com:tacman/twig-tree-tag-demo && cd twig-tree-tag-demo
composer install
symfony server:start -d
symfony open:local
```

It's a trival app -- install the library, register the extesion in services.yaml, and create a controller / template.  

Did something change in twig regarding output?  I vaguely remember something about twig and using the output buffer in a discussion.

It's a cool tag, it really simplifies rendering trees by skipping creating a macro to make recursive calls.  I didn't write it, but I did upgrade it a few years ago to Twig 3, and I'm certain at some point the twig 3 version worked.

Thanks.

**Comment by tacman** [comment]
@kbond Since you're a twig guru, can you point me on where to look?  I'm not even sure how to debug this.  Thx.

**Comment by stof** [comment]
Twig has a new yield-based implementation replacing the implementation based on output buffering (which will make it compatible with Fiber-based code as it avoids the global output buffering that would be shared between fibers).
Twig has a BC layer for non-yield-ready nodes. But maybe this BC layer fails to automatically handle your TreeNode logic.

**Comment by tacman** [comment]
Thanks, @stof .  Is there documentation for that yet?  Or a relatively simple example I can follow?  I've read https://twig.symfony.com/doc/3.x/advanced.html#defining-a-node but I can't see what I'm doing differently.

Or is it something that can be fixed in the BC layer?

**Comment by fabpot** [comment]
Here is the patch you need to apply to make it work:

```diff
diff --git a/src/Twig/Node/TreeNode.php b/src/Twig/Node/TreeNode.php
index 4f5fb0d..c637bdd 100644
--- a/src/Twig/Node/TreeNode.php
+++ b/src/Twig/Node/TreeNode.php
@@ -131,7 +131,7 @@ class TreeNode extends Node
         $compiler
             ->outdent()
             ->write("};\n")
-            ->write("\$tree_")
+            ->write("yield from \$tree_")
             ->raw($this->getAttribute('as'))
             ->raw("(")
             ->subcompile($this->getNode('seq'))
```

You will also need to remove the usage of the `twig_ensure_traversable` function as it's an internal function that has been removed.

The new yield implementation BC layer has many drawbacks that were not anticipated. I need to come back to the drawing board to see how we can mitigate these problems.

**Comment by tacman** [comment]
Thanks, it's working now!

I wasn't sure how to remove the call to twig_ensure_traversable, so I replaced it with the internal call that's used elsewhere:

```php
            ->write("\$context['_seq'] = CoreExtension::ensureTraversable(\$data);\n")
```

Is there a better way?

**Comment by fabpot** [comment]
> ```
> ensureTraversable
> ```

That's the way to go indeed.

**Comment by tacman** [comment]
Seems to be working fine now, thanks!

**Comment by ericmorand** [comment]
@fabpot

> That's the way to go indeed.

`CoreExtension::ensureTraversable` is internal too. How can this be the way to go if using `twig_ensure_traversable`, also internal, was _not_ the way to go and was removed without notice - I quote you:

> You will also need to remove the usage of the twig_ensure_traversable function as it's an internal function that has been removed.

**Comment by fabpot** [comment]
> @fabpot
> 
> > That's the way to go indeed.
> 
> `CoreExtension::ensureTraversable` is internal too. How can this be the way to go if using `twig_ensure_traversable`, also internal, was _not_ the way to go and was removed without notice - I quote you:
> 
> > You will also need to remove the usage of the twig_ensure_traversable function as it's an internal function that has been removed.

"The way to go" just means that `CoreExtension::ensureTraversable` is the replacement for `twig_ensure_traversable`.But to make it clear, this is still internal and not recommended for usage in user space.

**Comment by ericmorand** [comment]
@fabpot fair enough. Is there a recommended way?

**Comment by fabpot** [comment]
> @fabpot fair enough. Is there a recommended way?

That's really a one liner: `return is_iterable($seq) ? $seq : [];`

**Comment by ericmorand** [comment]
Fair enough, again. But why is this function internal if it is guaranteed that it will always resolve to the algorithm that you mention? I think you set as internal the methods that you don't want to consider as a public API because you don't want to face the cost of changing its contract.

What you are currently saying is that one can always use the mentioned algorithm to "ensure iterable", but at the same time you are not confident enough that it will always be the case to make the helper public.

Why are you not confident? Do you plan to change it in the future? If so, how can third party extensions be guaranteed that what they consider as iterable is the same thing as what the core extension consider as iterable?

**Comment by fabpot** [comment]
@tacman The `ensureTraversable()` method will disappear in Twig 4.0.

**Comment by tacman** [comment]
Where should the new way go?  

```php
            return is_iterable($seq) ? $seq : [];

//            ->write("\$context['_seq'] = twig_ensure_traversable(\$data);\n")
            ->write("\$context['_seq'] = CoreExtension::ensureTraversable(\$data);\n")
```

Should it be done like this?
  
```php
            ->write("\$context['_seq'] = is_iterable($data) ? $data : [];\n")
```

**Comment by fabpot** [comment]
> Where should the new way go?
> 
> ```
>             return is_iterable($seq) ? $seq : [];
> 
> //            ->write("\$context['_seq'] = twig_ensure_traversable(\$data);\n")
>             ->write("\$context['_seq'] = CoreExtension::ensureTraversable(\$data);\n")
> ```
> 
> Should it be done like this?
> 
> ```
>             ->write("\$context['_seq'] = is_iterable($data) ? $data : [];\n")
> ```

Yes!

**Comment by tacman** [comment]
This works:

```php
            // per https://github.com/twigphp/Twig/issues/4110
            ->write("\$context['_seq'] = is_iterable(\$data) ? \$data : [];\n")
```

I add "twig/twig:^3|^4" to composer, is there a way to test it with Twig 4?

**Comment by stof** [comment]
@tacman there is a 4.x-dev version on Packagist. You will have to configure composer to allow the dev stability.

And note that 4.x is an in-progress work right now. You should not make _releases_ that advocate support for Twig 4 yet (as the 4.x-dev branch does not reflect the final state of 4.x yet, so being compatible today does not mean you will be compatible with 4.0)

**Comment by tacman** [comment]
Thanks.  I think I've installed it correctly, and the tree works, but there's a deprecation error.  When I go to the debugger to see what it is, I get

> Cannot assign Symfony\Component\VarDumper\Cloner\Data to property Twig\Runtime\Loop::$seq of type Iterator"

Th line I changed, but it could also be related to the yield statements above.

```php
->write("\$context['_seq'] = is_iterable(\$data) ? \$data : [];\n")
```


```
An exception has been thrown during the rendering of a template ("Cannot assign Symfony\Component\VarDumper\Cloner\Data to property Twig\Runtime\Loop::$seq of type Iterator").
Twig\Error\
RuntimeError
Show exception properties
in vendor/symfony/web-profiler-bundle/Resources/views/Collector/logger.html.twig (line 516)
in [var/cache/dev/twig/4c/4c1110c5dbe44e789a03d9a1bf20ad14.php ](file:///home/tac/g/tacman/twig-tree-tag-demo/var/cache/dev/twig/4c/4c1110c5dbe44e789a03d9a1bf20ad14.php#L103)-> yieldBlock (line 103)
```

**Comment by tacman** [comment]
The demo repo is at https://github.com/tacman/twig-tree-tag-demo/tree/twig-4

**Comment by fabpot** [comment]
> Thanks. I think I've installed it correctly, and the tree works, but there's a deprecation error. When I go to the debugger to see what it is, I get
> 
> > Cannot assign Symfony\Component\VarDumper\Cloner\Data to property Twig\Runtime\Loop::$seq of type Iterator"
> 
> Th line I changed, but it could also be related to the yield statements above.
> 
> ```
> ->write("\$context['_seq'] = is_iterable(\$data) ? \$data : [];\n")
> ```
> 
> ```
> An exception has been thrown during the rendering of a template ("Cannot assign Symfony\Component\VarDumper\Cloner\Data to property Twig\Runtime\Loop::$seq of type Iterator").
> Twig\Error\
> RuntimeError
> Show exception properties
> in vendor/symfony/web-profiler-bundle/Resources/views/Collector/logger.html.twig (line 516)
> in [var/cache/dev/twig/4c/4c1110c5dbe44e789a03d9a1bf20ad14.php ](file:///home/tac/g/tacman/twig-tree-tag-demo/var/cache/dev/twig/4c/4c1110c5dbe44e789a03d9a1bf20ad14.php#L103)-> yieldBlock (line 103)
> ```

Good catch, fixed by #4141

**Comment by tacman** [comment]
I've updated my demo and everything's working, thanks!
